### PR TITLE
Use db.Find instead of writing methods for every object

### PR DESCRIPTION
**Pull request by lunny**
For those simple objects, it's unnecessary to write the find and count methods again and again.

### CODE REVIEW

File: routers/web/shared/actions/runners.go [hunk 1]
```diff
@@ -17,18 +17,13 @@ import (
 
 // RunnersList prepares data for runners list
 func RunnersList(ctx *context.Context, opts actions_model.FindRunnerOptions) {
-	count, err := actions_model.CountRunners(ctx, opts)
+	runners, count, err := db.FindAndCount[*actions_model.ActionRunner](ctx, opts)
```
Comment (delvh) [inline routers/web/shared/actions/runners.go]: Hey, even a performance improvement :grin:

File: routers/api/v1/org/org.go [hunk 2]
```diff
@@ -30,12 +30,12 @@ func listUserOrgs(ctx *context.APIContext, u *user_model.User) {
...
 		IncludePrivate: showPrivate,
 	}
-	orgs, err := organization.FindOrgs(ctx, opts)
+	orgs, err := db.Find[*organization.Organization](ctx, opts)
 	if err != nil {
 		ctx.Error(http.StatusInternalServerError, "FindOrgs", err)
 		return
 	}
-	maxResults, err := organization.CountOrgs(ctx, opts)
+	maxResults, err := db.Count[organization.Organization](ctx, opts)
```
Comment (delvh) [inline routers/api/v1/org/org.go]: Isn't that `FindAndCount` too?

File: routers/api/v1/notify/user.go [hunk 3]
```diff
@@ -68,18 +69,18 @@ func ListNotifications(ctx *context.APIContext) {
...
 
-	totalCount, err := activities_model.CountNotifications(ctx, opts)
+	totalCount, err := db.Count[activities_model.Notification](ctx, opts)
 	if err != nil {
 		ctx.InternalServerError(err)
 		return
 	}
 
-	nl, err := activities_model.GetNotifications(ctx, opts)
+	nl, err := db.Find[*activities_model.Notification](ctx, opts)
```
Comment (delvh) [inline routers/api/v1/notify/user.go]: Again, `FindAndCount`

File: models/auth/source.go [hunk 4]
```diff
@@ -257,19 +258,13 @@ func (opts FindSourcesOptions) ToConds() builder.Cond {
...
-}
-
 // IsSSPIEnabled returns true if there is at least one activated login
 // source of type LoginSSPI
 func IsSSPIEnabled(ctx context.Context) bool {
 	if !db.HasEngine {
 		return false
 	}
-	sources, err := FindSources(ctx, FindSourcesOptions{
+	sources, err := db.Find[Source](ctx, FindSourcesOptions{
```
Comment (delvh) [inline models/auth/source.go]: Isn't that an `db.Exists` (assuming we already have it)?

File: routers/web/user/notification.go [hunk 5]
```diff
@@ -409,5 +415,15 @@ func NotificationWatching(ctx *context.Context) {
...
 // NewAvailable returns the notification counts
 func NewAvailable(ctx *context.Context) {
-	ctx.JSON(http.StatusOK, structs.NotificationCount{New: activities_model.CountUnread(ctx, ctx.Doer.ID)})
+	total, err := db.Count[activities_model.Notification](ctx, activities_model.FindNotificationOptions{
+		UserID: ctx.Doer.ID,
+		Status: []activities_model.NotificationStatus{activities_model.NotificationStatusUnread},
+	})
+	if err != nil {
+		log.Error("db.Count[activities_model.Notification]", err)
+		ctx.JSON(http.StatusOK, structs.NotificationCount{New: 0})
```
Comment (delvh) [inline routers/web/user/notification.go:433]: :raised_eyebrow:

File: models/activities/notification_list.go [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+		sess = db.SetSessionPagination(sess, opts)
+	}
+	return sess
+}
+
+func getNotificationsByIssueID(ctx context.Context, issueID int64) (notifications []*Notification, err error) {
+	err = db.GetEngine(ctx).
+		Where("issue_id = ?", issueID).
+		Find(&notifications)
+	return notifications, err
```
Comment (delvh) [inline models/activities/notification_list.go]: Isn't that `db.Find` too?

File: models/activities/notification_list.go [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+		Where("user_id = ?", user.ID).
+		In("status", statuses).
+		OrderBy("updated_unix DESC")
+
+	if page > 0 && perPage > 0 {
+		sess.Limit(perPage, (page-1)*perPage)
+	}
+
+	err = sess.Find(&notifications)
+	return notifications, err
```
Comment (delvh) [inline models/activities/notification_list.go]: Again, `db.Find`?

File: models/activities/notification_list.go [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+}
+
+// NotificationList contains a list of notifications
+type NotificationList []*Notification
+
+// LoadAttributes load Repo Issue User and Comment if not loaded
+func (nl NotificationList) LoadAttributes(ctx context.Context) error {
+	var err error
+	for i := 0; i < len(nl); i++ {
+		err = nl[i].LoadAttributes(ctx)
```
Comment (delvh) [inline models/activities/notification_list.go]:
```suggestion
	for _, notification := range nl {
		err = notification.LoadAttributes(ctx)
```

File: models/activities/notification_list.go [hunk 9]
```diff
@@ -0,0 +1,362 @@
...
+				return nil, err
+			}
+
+			comments[comment.ID] = &comment
+		}
+		_ = rows.Close()
+
+		left -= limit
+		commentIDs = commentIDs[limit:]
+	}
```
Comment (delvh) [inline models/activities/notification_list.go]: I really dislike that we are defining the same method four times in this file for a different session.
Can't we unify them into a common function with sess as parameter?

Comment (lunny) [reply models/activities/notification_list.go]: I also want to unify them but I haven't found a good idea.

File: routers/web/user/notification.go (hunk 5 above)
Comment (lunny) [reply routers/web/user/notification.go:433]: ?

Comment (delvh) [reply routers/web/user/notification.go:433]: I think it's weird to respond with `OK` even though actually an error occurred.
Is that really how it should behave?

File: models/db/context.go [hunk 10]
```diff
@@ -264,3 +264,8 @@ func inTransaction(ctx context.Context) (*xorm.Session, bool) {
 		return nil, false
 	}
 }
+
+func Exits[T any](ctx context.Context, opts FindOptions) (bool, error) {
```
Comment (delvh) [inline models/db/context.go]: Typo
```suggestion
func Exists[T any](ctx context.Context, opts FindOptions) (bool, error) {
```

File: routers/web/user/notification.go (hunk 5 above)
Comment (lunny) [reply routers/web/user/notification.go:433]: I don't know, just followed previous logic.
If we want to change this, we need also change the frontend logic. Maybe leave it to other PRs.

**Review by delvh — APPROVED** [review]

File: models/activities/notification_test.go
Comment (nanguanlin6) [inline models/activities/notification_test.go:1]: those two tests can be converted as below(maybe the function names should be changed). No need to delete?
```
func TestNotificationsForUser(t *testing.T) {
	assert.NoError(t, unittest.PrepareTestDatabase())
	user := unittest.AssertExistsAndLoadBean(t, &user_model.User{ID: 2})
	notfs, err := db.Find[activities_model.Notification](db.DefaultContext, activities_model.FindNotificationOptions{
		UserID: user.ID,
		Status: []activities_model.NotificationStatus{
			activities_model.NotificationStatusRead,
			activities_model.NotificationStatusUnread,
		},
	})
	assert.NoError(t, err)
	if assert.Len(t, notfs, 3) {
		assert.EqualValues(t, 5, notfs[0].ID)
		assert.EqualValues(t, user.ID, notfs[0].UserID)
		assert.EqualValues(t, 4, notfs[1].ID)
		assert.EqualValues(t, user.ID, notfs[1].UserID)
		assert.EqualValues(t, 2, notfs[2].ID)
		assert.EqualValues(t, user.ID, notfs[2].UserID)
	}
}

func TestGetNotificationCount(t *testing.T) {
	assert.NoError(t, unittest.PrepareTestDatabase())
	user := unittest.AssertExistsAndLoadBean(t, &user_model.User{ID: 1})
	cnt, err := db.Count[activities_model.Notification](db.DefaultContext, activities_model.FindNotificationOptions{
		UserID: user.ID,
		Status: []activities_model.NotificationStatus{
			activities_model.NotificationStatusRead,
		},
	})
	assert.NoError(t, err)
	assert.EqualValues(t, 0, cnt)

	cnt, err = db.Count[activities_model.Notification](db.DefaultContext, activities_model.FindNotificationOptions{
		UserID: user.ID,
		Status: []activities_model.NotificationStatus{
			activities_model.NotificationStatusUnread,
		},
	})
	assert.NoError(t, err)
	assert.EqualValues(t, 1, cnt)
}
```

File: routers/api/v1/user/key.go [hunk 11]
```diff
@@ -56,25 +57,34 @@ func listPublicKeys(ctx *context.APIContext, user *user_model.User) {
...
+		})
 		if err2 != nil {
 			ctx.InternalServerError(err)
 			return
 		}
 		count = int(total)
 
 		// Use ListPublicKeys
-		keys, err = asymkey_model.ListPublicKeys(ctx, user.ID, utils.GetListOptions(ctx))
+		keys, err = db.Find[*asymkey_model.PublicKey](ctx, asymkey_model.FindPublicKeyOptions{
```
Comment (delvh) [inline routers/api/v1/user/key.go]: Isn't this `FindAndCount` too?

File: routers/api/v1/user/key.go [hunk 12]
```diff
@@ -56,25 +57,34 @@ func listPublicKeys(ctx *context.APIContext, user *user_model.User) {
...
-		} else {
-			// Unrestricted
-			keys, err = asymkey_model.SearchPublicKey(ctx, 0, fingerprint)
+			userID = user.ID
 		}
+		keys, err = db.Find[*asymkey_model.PublicKey](ctx, asymkey_model.FindPublicKeyOptions{
+			OwnerID:     userID,
+			Fingerprint: fingerprint,
+		})
 		count = len(keys)
```
Comment (delvh) [inline routers/api/v1/user/key.go]: `FindAndCount`

**Review by delvh — APPROVED** [review]

**Comment by wxiaoguang** [comment]
Why sometimes `db.Find[T]` but sometimes `db.Find[*T]`? Could there be a consistent behavior which only accepts `db.Find[T]` and returns `*T[]`?

File: models/auth/oauth2_grant.go [hunk 13]
```diff
@@ -0,0 +1,217 @@
+// Copyright 2023 The Gitea Authors. All rights reserved.
```
Comment (wxiaoguang) [inline models/auth/oauth2_grant.go]: Is this "code moving" related to the `db.FInd`? If it is not related, it just makes the PR more complex than it should be and may introduce new bugs.

**Comment by lunny** [comment]
> Why sometimes `db.Find[T]` but sometimes `db.Find[*T]`? Could there be a consistent behavior which only accepts `db.Find[T]` and returns `*T[]`?

All db.Find and db.FindAndCount has been using pointer to struct but for db.Count, if we use a pointer, we need introduce reflect so I don't think it's necessary. So all db.Count should use struct directly.

File: models/auth/oauth2_grant.go (hunk 13 above)
Comment (lunny) [reply models/auth/oauth2_grant.go]: Done.

**Comment by wxiaoguang** [comment]
How to avoid low-level mistakes? When reviewing PRs, which is right while which is wrong?

* db.Find[T]
* db.Find[*T]
* db.Count[T]
* db.Count[*T]


I do not think leaving low-level mistakes in a framework is healthy to a big project.

**Comment by lunny** [comment]
> How to avoid low-level mistakes? When reviewing PRs, which is right while which is wrong?
> 
>     * db.Find[T]
> 
>     * db.Find[*T]
> 
>     * db.Count[T]
> 
>     * db.Count[*T]
> 
> 
> I do not think leaving low-level mistakes in a framework is healthy to a big project.

[x] db.Find[T]
[x] db.Find[*T]
[x] db.Count[T]
x  db.Count[*T]

Let me try to avoid `db.Count[*T]`.

**Comment by wxiaoguang** [comment]
> [x] db.Find[T]
> [x] db.Find[*T]
> [x] db.Count[T]

Why not just use `db.Find[T](...) *T[]` and `db.Count[T](...)`?

**Review by wxiaoguang — APPROVED** [review]

**Pull request merged**